### PR TITLE
Make `SeaHash` compatible with `hashlib.file_digest`

### LINKED ISSUE #24: `SeaHash` is not compatible with `hashlib.file_digest`
The following snippet:
```python
    with open("file.txt", "rb") as f:
        s = hashlib.file_digest(f, lambda: seahash.SeaHash())
```
fails with the following error:
```
>           digestobj.update(view[:size])
E           TypeError: argument 'obj': 'memoryview' object cannot be converted to 'PyBytes'

C:\Python312\Lib\hashlib.py:236: TypeError
```

### PR DESCRIPTION
**Pull request by tlaferriere**
Fixes #24 

I used the python buffer protocol to copy the bytes out of a bytes-like object passed to the function if it doesn't convert to `&[u8]`.
This introduces a copy of the buffer, so the performance may not be on-par with what the seahash library can do, but at least it doesn't outright break when using the `hashlib.file_digest` function with our hash class.
On the other hand, instead of appending bytes to an inner vec, I used the streaming hasher to hash the blocks as they come in, removing a copy and reducing total memory usage from O(n) to O(1) relative to the object-being-hashed's size.

- [x] Make `SeaHash.update()` accept memoryview objects

### CODE REVIEW

File: seahash.pyi [hunk 1]
```diff
@@ -6,7 +8,7 @@ class SeaHash:
     block_size: int
 
     def __init__(self, string: bytes = b""): ...
```
Comment (RealOrangeOne) [inline seahash.pyi:10]:
```suggestion
    def __init__(self, data: bytes = b""): ...
```

File: src/lib.rs [hunk 2]
```diff
@@ -31,25 +36,41 @@ mod inner {
...
 
         pub fn digest(&self, py: Python) -> PyObject {
-            PyBytes::new(py, &self.intdigest(py).to_ne_bytes()).into()
+            PyBytes::new(py, &self.intdigest().to_ne_bytes()).into()
         }
 
-        pub fn intdigest(&self, py: Python) -> u64 {
-            hash(py, &self.inner)
+        pub fn intdigest(&self) -> u64 {
+            self.inner.finish()
```
Comment (RealOrangeOne) [inline src/lib.rs:69]: Suggestion: Given `finish` likely computes the final hash, should it unlock the GIL?

File: seahash.pyi [hunk 3]
```diff
@@ -1,3 +1,5 @@
+from collections.abc import Buffer
```
Comment (RealOrangeOne) [inline seahash.pyi:1]: Issue: `Buffer` only exists in 3.12, so this won't work on older versions. I think just using `memoryview` is fine here.

File: src/lib.rs [hunk 4]
```diff
@@ -31,25 +36,41 @@ mod inner {
...
-            self.inner.extend_from_slice(obj)
+            self.inner.write(obj)
+        }
+
+        #[cfg(Py_3_11)]
+        pub fn update(&mut self, py: Python, obj: &PyAny) -> PyResult<()> {
+            match obj.extract() {
+                Ok(buf) => {
+                    self.inner.write(buf);
+                }
```
Comment (RealOrangeOne) [inline src/lib.rs:56]: Suggestion: This is likely better written using `if let Ok(buf) = obj.extract()`.

File: tests/test_seahash.py [hunk 5]
```diff
@@ -29,6 +30,23 @@ def test_hashlib_compatible():
     assert s.intdigest() == int.from_bytes(s.digest(), sys.byteorder)
 
 
+@pytest.mark.skipif(
+    sys.version_info < (3, 11),
+    reason="requires hashlib.file_digest which is new in python3.11",
+)
+def test_hashlib_file_digest_compatible(tmp_path):
+    import hashlib
```
Comment (RealOrangeOne) [inline tests/test_seahash.py:38]: Nit: This import should live at the top of the file. I realise it's only useful for this test, but it's still easier to maintain if imports are done in 1 place.

File: tests/test_seahash.py [hunk 6]
```diff
@@ -29,6 +30,23 @@ def test_hashlib_compatible():
...
+
+    p = tmp_path / "hello.txt"
+    CONTENT = b"Hash me!"
```
Comment (RealOrangeOne) [inline tests/test_seahash.py:41]: Nit: As `content` is not a constant, it should be `lower_case`.

File: src/lib.rs [hunk 7]
```diff
@@ -31,25 +36,41 @@ mod inner {
...
+                Err(_) => {
+                    self.inner.write(PyBuffer::get(obj)?.to_vec(py)?.as_slice());
```
Comment (RealOrangeOne) [inline src/lib.rs:58]: Question: Could we use overloading instead to get the correct type for this? There's 2 fairly separate implementations here.

File: src/lib.rs (hunk 2 above)
Comment (tlaferriere) [reply src/lib.rs:69]: The hash gets updated every time bytes are written to the hasher using `write` (which calls out to [`push_bytes`](https://docs.rs/seahash/latest/src/seahash/stream.rs.html#59-157)). Technically, `update` would gain more by unlocking the GIL. [`finish`](https://docs.rs/seahash/latest/src/seahash/stream.rs.html#166-175) only computes the hash of the tail if there is one.

File: seahash.pyi (hunk 3 above)
Comment (tlaferriere) [reply seahash.pyi:1]: Good catch!

File: src/lib.rs (hunk 7 above)
Comment (tlaferriere) [reply src/lib.rs:58]: Rust doesn't support function overloading... Maybe PyO3 can help out here? I'm not sure how to achieve this.

**Pull request merged**